File: interpreter/src/main/java/cambio/tltea/interpreter/nodes/TriggerNotifier.java
package cambio.tltea.interpreter.nodes;

import cambio.tltea.interpreter.nodes.cause.EventActivationListener;
import cambio.tltea.interpreter.nodes.cause.EventActivationNode;
import org.jetbrains.annotations.NotNull;

import java.util.Collection;
import java.util.HashSet;
import java.util.List;

/**
 * @author Lion Wagner
 */
public class TriggerNotifier {

    protected final Collection<ITriggerListener> subscribers = new HashSet<>();

    public final void subscribe(ITriggerListener listener) {
        subscribers.add(listener);
    }

    public final void unsubscribe(ITriggerListener listener) {
        subscribers.remove(listener);
    }

    public void trigger(String eventName, Object... args) {
        for (ITriggerListener listener : subscribers) {
            listener.onTrigger(eventName, args);
        }
    }

    public void activateListeners(@NotNull List<EventActivationListener> listeners) {

    }
}

File: interpreter/src/main/java/cambio/tltea/interpreter/nodes/TriggerNotifier.kt
package cambio.tltea.interpreter.nodes

import cambio.tltea.interpreter.nodes.consequence.ActivationData
import java.util.function.Consumer

/**
 * @author Lion Wagner
 */
class TriggerNotifier : ISubscribableTriggerNotifier {

    private val anySubscribers: MutableCollection<Consumer<ActivationData<*>>> = HashSet()

    private val filteredSubscribers: HashMap<Class<ActivationData<*>>, MutableSet<Consumer<ActivationData<*>>>> =
        HashMap()

    override fun subscribeEventListener(listener: Consumer<ActivationData<*>>) {
        anySubscribers.add(listener)
    }

    override fun subscribeEventListenerWithFilter(listener: Consumer<ActivationData<*>>, filter: Class<ActivationData<*>>) {
        if (!filteredSubscribers.containsKey(filter)) {
            filteredSubscribers[filter] = HashSet()
        }
        filteredSubscribers[filter]!!.add(listener)
    }

    override fun unsubscribe(listener: Consumer<ActivationData<*>>) {
        anySubscribers.remove(listener)
        filteredSubscribers.values.forEach { it.remove(listener) }
    }

    internal fun trigger(activationData: ActivationData<*>) {
        anySubscribers.forEach { it.accept(activationData) }
        filteredSubscribers[activationData.javaClass]?.forEach { it.accept(activationData) }
    }
}